libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -292,46 +276,10 @@ bool TypeInference::visit(InlineAssembly const& _inlineAssembly)
 	return false;
 }
 
-bool TypeInference::visit(ElementaryTypeNameExpression const& _expression)
+bool TypeInference::visit(BuiltinDefinition const&)
 {
-	auto& expressionAnnotation = annotation(_expression);
-	solAssert(!expressionAnnotation.type);
-
-	switch (m_expressionContext)
-	{
-	case ExpressionContext::Term:
-	case ExpressionContext::Type:
-		if (auto constructor = m_analysis.annotation<TypeRegistration>(_expression).typeConstructor)
-		{
-			std::vector<Type> arguments;
-			std::generate_n(std::back_inserter(arguments), m_typeSystem.constructorInfo(*constructor).arguments(), [&]() {
-				return m_typeSystem.freshTypeVariable({});
-			});
-			// TODO: get rid of the distinction (assign a function on unit for the empty case)? Ambiguity?
-			if (arguments.empty() || m_expressionContext == ExpressionContext::Term)
-				expressionAnnotation.type = m_typeSystem.type(*constructor, arguments);
-			else
-			{
-				TypeSystemHelpers helper{m_typeSystem};
-				expressionAnnotation.type =
-					helper.typeFunctionType(
-						helper.tupleType(arguments),
-						m_typeSystem.type(*constructor, arguments)
-					);
```

Comment:
Do I still need this? I suspect I should now be doing this in `visit(Identifier)` when the identifier is a type name, but (at least in my simple tests) everything seems to be working just fine without it.

But maybe that's because I need more complex examples? Some expressions on types in variable declarations? Are those even implemented?

Reply:
All this becoming obsolete is kind of the point in making these types defined as usual - so it should work fine (and if it doesn't yet in all cases, we'll fix that elsewhere)

Reply:
But I meant this specific part of the function that I highlighted above. The assignment of a type annotation in term context is indeed handled when visiting the type definition now, but it does nothing in type context.

Reply:
I'm not sure I follow exactly :-). There is (may be) *some* difference between the old handling as ``ElementaryTypeNameExpression``s and the new one based on ``Identifier``s that refer to a type declaration (which are handled in ``handleIdentifierByReferencedDeclaration``).

But I think the new handling in ``Identifier`` should be the correct one. The old version may not even have worked for "function calls" to elementary type names in type expressions (not sure that even parsed), while the new one should.

In any case (without digging into it completely myself now), I think this is fine and if we notice that anything is missing later on, we can *then* fix it in the identifier visit.